File: vaultpass_backend/src/vaultpass_backend/models/document.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, List
from sqlalchemy import String, Text, DateTime, ForeignKey, Enum as SQLEnum, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from vaultpass_backend.database.connection import Base

if TYPE_CHECKING:
    from vaultpass_backend.models.user import User
    from vaultpass_backend.models.document_share import DocumentShare

class DocumentType(str, enum.Enum):
    PASSPORT = "PASSPORT"
    NATIONAL_ID = "NATIONAL_ID"
    BIRTH_CERTIFICATE = "BIRTH_CERTIFICATE"
    ACADEMIC_CERTIFICATE = "ACADEMIC_CERTIFICATE"
    PROPERTY_DOCUMENT = "PROPERTY_DOCUMENT"
    WILL = "WILL"
    OTHER = "OTHER"

class Document(Base):
    """
    SQLAlchemy model representing the documents table.
    """
    __tablename__ = "documents"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        default=uuid.uuid4,
        index=True
    )
    owner_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    title: Mapped[str] = mapped_column(
        String(255),
        nullable=False
    )
    document_type: Mapped[DocumentType] = mapped_column(
        SQLEnum(DocumentType, name="document_type_enum"),
        nullable=False
    )
    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True
    )
    file_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False
    )
    file_path: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )
    file_size: Mapped[int] = mapped_column(
        nullable=False
    )
    mime_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False
    )
    expiry_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True
    )
    uploaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        onupdate=func.now(),
        nullable=True
    )

    # Relationships
    owner: Mapped["User"] = relationship("User", back_populates="documents")
    shares: Mapped[List["DocumentShare"]] = relationship(
        "DocumentShare",
        back_populates="document",
        cascade="all, delete-orphan"
    )

File: frontend/src/types/document.ts
export type DocumentType =
  | 'PASSPORT'
  | 'NATIONAL_ID'
  | 'BIRTH_CERTIFICATE'
  | 'ACADEMIC_CERTIFICATE'
  | 'PROPERTY_DOCUMENT'
  | 'WILL'
  | 'OTHER';

export type DocumentCategoryFilter = 'ALL' | 'ID' | 'LEGAL' | 'MEDICAL';

export interface DocumentModel {
  id: string;
  owner_id: string;
  title: string;
  document_type: DocumentType;
  description?: string | null;
  file_name: string;
  file_path: string;
  file_size: number;
  mime_type: string;
  expiry_date?: string | null;
  uploaded_at: string;
  created_at: string;
  updated_at?: string | null;
}

export interface DocumentListResponse {
  items: DocumentModel[];
  total: number;
  page: number;
  limit: number;
}

export interface DocumentUploadPayload {
  file: File;
  title: string;
  document_type: DocumentType;
  description?: string;
  expiry_date?: string;
}

export interface DocumentUpdatePayload {
  title: string;
  document_type: DocumentType;
  description?: string;
  expiry_date?: string;
}

export const DOCUMENT_TYPE_LABELS: Record<DocumentType, string> = {
  PASSPORT: 'Passport',
  NATIONAL_ID: 'National ID',
  BIRTH_CERTIFICATE: 'Birth Certificate',
  ACADEMIC_CERTIFICATE: 'Academic Certificate',
  PROPERTY_DOCUMENT: 'Property Document',
  WILL: 'Last Will',
  OTHER: 'Other',
};

export const DOCUMENT_TYPE_ICONS: Record<DocumentType, string> = {
  PASSPORT: 'badge',
  NATIONAL_ID: 'badge',
  BIRTH_CERTIFICATE: 'badge',
  ACADEMIC_CERTIFICATE: 'school',
  PROPERTY_DOCUMENT: 'gavel',
  WILL: 'gavel',
  OTHER: 'description',
};

export const DOCUMENT_CATEGORY_BUTTONS: ReadonlyArray<{
  key: DocumentCategoryFilter;
  label: string;
}> = [
  { key: 'ALL', label: 'All Documents' },
  { key: 'ID', label: 'ID' },
  { key: 'LEGAL', label: 'Legal' },
  { key: 'MEDICAL', label: 'Medical' },
];

export const DOCUMENT_CATEGORY_TYPE_MAP: Record<DocumentCategoryFilter, DocumentType[]> = {
  ALL: [
    'PASSPORT',
    'NATIONAL_ID',
    'BIRTH_CERTIFICATE',
    'ACADEMIC_CERTIFICATE',
    'PROPERTY_DOCUMENT',
    'WILL',
    'OTHER',
  ],
  ID: ['PASSPORT', 'NATIONAL_ID', 'BIRTH_CERTIFICATE'],
  LEGAL: ['PROPERTY_DOCUMENT', 'WILL'],
  MEDICAL: [],
};
